### PR TITLE
Add UTXO Global Provider

### PR DESCRIPTION
**Pull request by trongdth**
## What?
- We added code to support the "utxoGlobal" provider, a new Chrome Wallet Extension that runs on the CKB network. More information: [utxoGlobal](https://utxo.global/)

## Why?
- These changes are fully compatible with version 0.0.1 on the Chrome Store.

## Testing?

https://github.com/user-attachments/assets/350bbcd0-28ed-41ba-b7e5-dcd4fffa23fb

## Anything Else?
- The UTXO Global Wallet v0.0.1 does not yet support the "switchNetwork" function. Therefore, you need to switch between Bitcoin or CKB manually.
- The "signTransaction" feature has not been implemented yet. It will be available in the next version and will be updated later on.

### CODE REVIEW

**Comment by trongdth** [comment]
[EDITED] 26 Jul 

- The new commit has supported the "switchNetwork" + "signTransaction" functions. This commit is compatible with version 0.0.2 of UTXO Global Wallet Extension (commit https://github.com/UTXO-Global/utxo-wallet-extension/commit/6a6ee495470b3af18ddf878144d5d2143cacf226)
- Version 0.0.2 has not yet been submitted to Chrome Store because version 0.0.1 is still pending.
- The video demo is attached

https://github.com/user-attachments/assets/7de2923a-f02f-4ff1-8a34-461d88a2126f


** Happy coding **

File: packages/connector/src/connector/index.ts [hunk 1]
```diff
@@ -173,7 +173,7 @@ export class WebComponentConnector extends LitElement {
       return;
     }
     this.mainRef.value.style.height = `${
-      this.bodyRef.value?.clientHeight ?? 0
+      this.bodyRef.value?.offsetHeight ?? 0
```
Comment (Hanssen0) [inline packages/connector/src/connector/index.ts:176]: Please revert this change

File: packages/connector/src/signers/index.ts [hunk 2]
```diff
@@ -90,6 +91,10 @@ export class SignersController implements ReactiveController {
       this.addSigner("Nostr", "Nostr", NOSTR_SVG, nip07Signer);
     }
 
+    ccc.UtxoGlobal.getUtxoGlobalSigners(client)?.forEach(({ signer, name }) => {
```
Comment (Hanssen0) [inline packages/connector/src/signers/index.ts:94]: Since it returns an array, it should always be not null. (But can be empty)

File: packages/utxo-global/src/btc/index.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  async getBtcAccount() {
+    const accounts = await this.provider.getAccount();
+    return accounts[0];
+  }
+
+  async getBtcPublicKey(): Promise<ccc.Hex> {
+    const pubKeys = await this.provider.getPublicKey();
+    const account = await this.getBtcAccount();
+    const pubKey = pubKeys.find((_pubKey) => _pubKey.address === account);
+    return ccc.hexFrom(pubKey?.publicKey!);
```
Comment (Hanssen0) [inline packages/utxo-global/src/btc/index.ts:21]: It's better to throw an error in unexpected situations when not forced by type explicitly.

File: packages/utxo-global/src/btc/index.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  async getBtcAccount() {
+    const accounts = await this.provider.getAccount();
+    return accounts[0];
+  }
+
+  async getBtcPublicKey(): Promise<ccc.Hex> {
+    const pubKeys = await this.provider.getPublicKey();
+    const account = await this.getBtcAccount();
+    const pubKey = pubKeys.find((_pubKey) => _pubKey.address === account);
```
Comment (Hanssen0) [inline packages/utxo-global/src/btc/index.ts:20]: _ prefix usually for not used variable but that's not the situation here.

File: packages/utxo-global/src/ckb/index.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+  async getAddressObj(): Promise<ccc.Address | undefined> {
+    const address = await this.getInternalAddress();
+    return await ccc.Address.fromString(address, this.client);
+  }
+
+  async getAddressObjs(): Promise<ccc.Address[]> {
+    const address = await this.getAddressObj();
+
+    if (!!address) {
```
Comment (Hanssen0) [inline packages/utxo-global/src/ckb/index.ts:47]: if the address is empty, which means the wallet is not correctly connected, we should throw an error instead of acting like normal.

File: packages/utxo-global/src/ckb/index.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+  ) {
+    super(client);
+  }
+
+  getInternalAddress(): Promise<string> {
+    return this.getAccount();
+  }
+
+  async getIdentity(): Promise<string> {
+    return (await this.getPublicKey()).slice(2);
```
Comment (Hanssen0) [inline packages/utxo-global/src/ckb/index.ts:36]: Please keep the 0x prefix. Consider extending the SignerCkbPublicKey?

File: packages/connector/src/signers/index.ts (hunk 2 above)
Comment (trongdth) [reply packages/connector/src/signers/index.ts]: fixed

File: packages/connector/src/connector/index.ts (hunk 1 above)
Comment (trongdth) [reply packages/connector/src/connector/index.ts]: fixed

File: packages/utxo-global/src/btc/index.ts (hunk 3 above)
Comment (trongdth) [reply packages/utxo-global/src/btc/index.ts]: fixed

File: packages/utxo-global/src/btc/index.ts (hunk 4 above)
Comment (trongdth) [reply packages/utxo-global/src/btc/index.ts]: fixed

File: packages/utxo-global/src/ckb/index.ts (hunk 5 above)
Comment (trongdth) [reply packages/utxo-global/src/ckb/index.ts:41]: fixed

**Comment by trongdth** [comment]
The feedbacks were solved. Could you please re-take a look at it? @Hanssen0

File: packages/core/package.json [hunk 7]
```diff
@@ -60,6 +60,9 @@
     "@joyid/ckb": "^1.0.1",
     "@noble/curves": "^1.4.2",
     "@noble/hashes": "^1.4.0",
+    "@types/secp256k1": "^4.0.6",
```
Comment (Hanssen0) [inline packages/core/package.json:63]: These deps should be removed now.

File: packages/utxo-global/README.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
+<p align="center">
+  <a href="https://ckbccc-demo.vercel.app/">
+    <img alt="Logo" src="https://raw.githubusercontent.com/ckb-ecofund/ccc/master/assets/logo.svg" width="20%" />
+  </a>
+</p>
+
+<h1 align="center" style="font-size: 64px;">
+  CCC's support for Utxo Global
```
Comment (Hanssen0) [inline packages/utxo-global/README.md:8]: It's OK to use UTXO in the README if you want.

File: packages/utxo-global/package.json [hunk 9]
```diff
@@ -0,0 +1,55 @@
+{
+  "name": "@ckb-ccc/utxo-global",
+  "version": "0.0.5-alpha.10",
```
Comment (Hanssen0) [inline packages/utxo-global/package.json:3]: Please modify the version to be the same as others.

File: packages/utxo-global/src/btc/index.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  async signMessageRaw(message: string | ccc.BytesLike): Promise<string> {
+    const challenge =
+      typeof message === "string" ? message : ccc.hexFrom(message).slice(2);
+    const account = await this.getBtcAccount();
+    return this.provider.signMessage(challenge, account);
+  }
+
+  async signMessage(message: ccc.BytesLike): Promise<ccc.Signature> {
```
Comment (Hanssen0) [inline packages/utxo-global/src/btc/index.ts:44]: You can remove this, it has a default implementation.

File: packages/utxo-global/src/ckb/index.ts [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  async signMessageRaw(message: string | ccc.BytesLike): Promise<string> {
+    const challenge =
+      typeof message === "string" ? message : ccc.hexFrom(message).slice(2);
+    const account = await this.getAccount();
+    return this.provider.signMessage(challenge, account);
+  }
+
+  async signMessage(message: ccc.BytesLike): Promise<ccc.Signature> {
```
Comment (Hanssen0) [inline packages/utxo-global/src/ckb/index.ts:80]: You can remove this, it has a default implementation.

And also verifyMessage and verifyMessageRaw below.

File: packages/utxo-global/src/ckb/index.ts [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    const pubKey = await this.getPublicKey();
+    return ccc.verifyMessageCkbSecp256k1(message, signature as string, pubKey);
+  }
+
+  async signOnlyTransaction(
+    txLike: ccc.TransactionLike,
+  ): Promise<ccc.Transaction> {
+    const rawTx = cccA.JsonRpcTransformers.transactionFrom(txLike);
+    const txSigned = await this.provider.signTransaction(rawTx);
+    return JSON.parse(txSigned) as ccc.Transaction;
```
Comment (Hanssen0) [inline packages/utxo-global/src/ckb/index.ts:121]: The ccc.Transaction has extra props like methods, use ccc.Transaction.from instead.

**Comment by Hanssen0** [comment]
Merged https://github.com/ckb-ecofund/ccc/commit/8b7ace26745d8ad21002d97d25bd859888da220c
